### DISCUSSION TITLE
Why another magic cli (gotk) instead of declarative description (code) ?

### DISCUSSION THREAD

**Discussion by vladimirfx** [fluxcd/flux2#248] [Q&A] [answered]
I've carefully read the manual and confused with big amount of cli calls needed to configure resources. Especially in comparison to FluxCD 1.x .
What a point in such drift from declarative to imperative approach with huge commands? (witch is very error prone, I've tried to 'compile' it myself)
Didn't GitOps means code, not cli? Ok, with minimal cluster bootstrap. I think I missing something... Can you clarify please?

PS: we are using FluxCD based on multi tenant sample and looking to new GitOps toolkit as Flux successor.

**Comment by stefanprodan** [comment, marked as answer]
The GitOps Toolkit is made out of Kubernetes manifests (for installation) and Kubernetes custom resources (for configuration). The CLI gives you the option to install and configure the toolkit with commands instead of YAMLs.

**Reply by vladimirfx** [reply]
Huh. Thank you a lot! Is there any guide/sample for declarative setup with minimal cli bootstrap?

**Reply by stefanprodan** [reply]
Check out the install guide https://toolkit.fluxcd.io/guides/installation/#generic-git-server 

PS. Any `gotk create` has an `--export` option that will write the YAML to stdout. You don't have to use the create command, instead write the YAMLs by hand and commit those to the repo. The `create` commands are there to help you get started and avoid typos, but none are required to run the toolkit.

**Reply by stefanprodan** [reply]
> we are using FluxCD based on multi tenant sample and looking to new GitOps toolkit as Flux successor.

After the [v0.1.0 release](https://github.com/fluxcd/toolkit/issues/238) we will publish a multi-tenancy guide for the toolkit.

**Reply by vladimirfx** [reply]
> Check out the install guide https://toolkit.fluxcd.io/guides/installation/#generic-git-server
> 
> PS. Any `gotk create` has an `--export` option that will write the YAML to stdout. You don't have to use the create command, instead write the YAMLs by hand and commit those to the repo. The `create` commands are there to help you get started and avoid typos, but none are required to run the toolkit.

That is minimal set of commands to bootstrap GitOps managed cluster is:
1. `gotk install` to create base infrastructure
2. `gotk create source git` to create git first source repo
3. `gotk create kustomization` to register first kustomisation from witch all other objects may be created declaratively.

All commands above may be used with `--export` flag  (i.e. offline) to generate Kubernates manifests witch may be applied with kubectl. After such setup gotk tool is not required anymore because other resources (sources, kustomizations, helm releases, etc) can be described through first kustomization resource defined in p. 3 in fully declarative way.

Correct me please if I'm wrong somewhere.

**Reply by stefanprodan** [reply]
Yes, if you prefer to do it all manually without gotk then:

1. `kustomize build github.com/fluxcd/toolkit/install?ref=v0.0.28 | kubectl apply -f-`
2. generate SSH keys and known_hosts
3. create a secret with the above and a GitRepository as described [here](https://toolkit.fluxcd.io/components/source/gitrepositories/#ssh-authentication)
4. set the deploy key in Git
5. apply the GitRepository with kubectl
6. create a  Kustomization YAML and apply it with kubectl
7. place both GitRepository and the Kustomization YAMLs in git
8. place any YAML in the repo and it will sync

**Reply by stefanprodan** [reply]
All the above steps can be automated for GitHub/GitLab with a single command: `gotk bootstrap`. It's up to you of you want to do it manually, but I think most users will prefer to run a simple command, that's why we made the gotk CLI.

**Reply by vladimirfx** [reply]
Thanks for such detailed response! Current Getting started guid slightly confused me by extensive usage of gotk tool. In our current workflow (Flux) most of team are not using special GitOps cli tools and someone even not known about it existence. Cluster usually bootstrapped by Ops team. Then it used declarative mostly way by changing configs in Git (copy-edit-verify-commit).

**Reply by Starefossen** [reply]
Second what @vladimirfx. We set up everything automatically and prefer to either use Terraform or from Git via Flux so having to run a `gotk` command does not fit very well to either of these two  workflows and creates a sort of arbitrary step where we have to run `gotk` at the end (or in the middle) of our pipeline for boostrapping new clusters.

**Reply by phillebaba** [reply]
One option may be to write a terraform provider as an alternative to the CLI. It might be a bit tricky building the state sync though as we would need to reverse what the CLI does.

**Reply by vladimirfx** [reply]
As I understand @stefanprodan all setup done by gotk can be reproduced by k8s manifests. IMO It can be deployed by terraform kustimoze provider. Our ops team are prototyping such setup.

**Reply by vladimirfx** [reply]
> Second what @vladimirfx. We set up everything automatically and prefer to either use Terraform or from Git via Flux so having to run a `gotk` command does not fit very well to either of these two workflows and creates a sort of arbitrary step where we have to run `gotk` at the end (or in the middle) of our pipeline for boostrapping new clusters.

We successfully bootstrapped several clusters by Terraform and Kustomize which is managed by gotk afterward. To do so we generate set of gotk manifests as documented in getting started guide for custom git repo. Then add simple kustomization that replaces gotk-system repository target and secrets. When cluster is bootstrapped by Terraform we apply gotk-system kustomization (with Terraform as well) and then we manage cluster by means of gotk (mostly).
